File: rust/src/mat8.rs
use std::io::{self, Write};
use std::thread;
use std::time::Duration;
use rand::Rng;

const COLS: usize = 80;
const ROWS: usize = 24;
const DELAY: u64 = 80000;

fn main() -> io::Result<()> {
    let mut rng = rand::thread_rng();
    
    // Initialize matrix
    let mut matrix = [[' '; COLS]; ROWS];
    let mut column_start = [0; COLS];
    let mut column_length = [0; COLS];
    
    // Random initial values
    for j in 0..COLS {
        column_start[j] = -(rng.gen::<i32>() % (ROWS as i32 * 2));
        column_length[j] = rng.gen_range(5..15);
    }
    
    // Hide cursor and clear screen
    print!("\x1B[?25l\x1B[2J");
    io::stdout().flush()?;
    
    // Handle Ctrl+C to restore cursor
    ctrlc::set_handler(move || {
        print!("\x1B[?25h");
        std::process::exit(0);
    }).expect("Error setting Ctrl-C handler");
    
    loop {
        print!("\x1B[H"); // Move cursor to home position
        
        // Update columns
        for j in 0..COLS {
            if column_start[j] > (ROWS as i32 + column_length[j] as i32) {
                if rng.gen_ratio(1, 20) {
                    column_start[j] = -(rng.gen::<i32>() % (ROWS as i32 * 2));
                    column_length[j] = rng.gen_range(5..15);
                }
            }
            
            if rng.gen_ratio(1, 3) && column_start[j] >= 0 {
                // Sometimes skip update
            } else {
                column_start[j] += 1;
            }
            
            // Clear position behind the column
            let clear_pos = column_start[j] - column_length[j] as i32 - 1;
            if clear_pos >= 0 && clear_pos < ROWS as i32 {
                matrix[clear_pos as usize][j] = ' ';
            }
            
            // Fill new positions
            let start = column_start[j];
            for i in start..start + column_length[j] as i32 {
                if i >= 0 && i < ROWS as i32 {
                    matrix[i as usize][j] = if rng.gen_ratio(6, 10) {
                        rng.gen_range(b'0'..=b'9') as char
                    } else {
                        rng.gen_range(b'A'..=b'Z') as char
                    };
                }
            }
        }
        
        // Display matrix
        for i in 0..ROWS {
            for j in 0..COLS {
                if matrix[i][j] != ' ' {
                    // Calculate brightness
                    let mut brightness = 0;
                    for k in 0..3 {
                        if i >= k && matrix[i - k][j] != ' ' {
                            brightness += 1;
                        }
                    }
                    
                    // Color based on brightness
                    match brightness {
                        1 => print!("\x1B[97m{}\x1B[0m", matrix[i][j]), // White
                        2 => print!("\x1B[92m{}\x1B[0m", matrix[i][j]), // Bright green
                        _ => print!("\x1B[32m{}\x1B[0m", matrix[i][j]), // Green
                    }
                    
                    // Randomly clear some characters
                    if rng.gen_ratio(1, 8) {
                        matrix[i][j] = ' ';
                    }
                } else {
                    print!(" ");
                }
            }
            println!();
        }
        
        io::stdout().flush()?;
        thread::sleep(Duration::from_micros(DELAY));
    }
}

File: mat8.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>

#define COLS 80
#define ROWS 24
#define DELAY 80000
#define MIN_START_DELAY 5
#define MAX_START_DELAY 15

int main() {
    srand(time(0));
    
    char matrix[ROWS][COLS];
    int column_start[COLS];
    int column_length[COLS];
    
    
    for (int i = 0; i < COLS; i++) {
        column_start[i] = -rand() % (ROWS * 2);
        column_length[i] = rand() % 10 + 5;
    }
    
    printf("\033[?25l");  
    printf("\033[2J");    
    
    while (1) {
        printf("\033[H"); 
        
        
        for (int j = 0; j < COLS; j++) {
        
            if (column_start[j] > ROWS + column_length[j]) {
                if (rand() % 20 == 0) {
                    column_start[j] = -rand() % (ROWS * 2);
                    column_length[j] = rand() % 10 + 5;
                }
            }
            
           
            if (rand() % 3 != 0 || column_start[j] < 0) {
                column_start[j]++;
            }
           
            int clear_pos = column_start[j] - column_length[j] - 1;
            if (clear_pos >= 0 && clear_pos < ROWS) {
                matrix[clear_pos][j] = ' ';
            }
            
            int start = column_start[j];
            for (int i = start; i < start + column_length[j]; i++) {
                if (i >= 0 && i < ROWS) {
                    matrix[i][j] = (rand() % 10 > 3) 
                        ? (rand() % 10 + '0') 
                        : (rand() % 26 + 'A');
                }
            }
        }
        
        
        for (int i = 0; i < ROWS; i++) {
            for (int j = 0; j < COLS; j++) {
                if (matrix[i][j] != ' ') {
                    
                    int brightness = 0;
                    for (int k = 0; k < 3; k++) {
                        if (i - k >= 0 && matrix[i - k][j] != ' ') {
                            brightness++;
                        }
                    }
                    
                    if (brightness == 1) {
                        printf("\033[97m%c\033[0m", matrix[i][j]);  
                    } else if (brightness == 2) {
                        printf("\033[92m%c\033[0m", matrix[i][j]);  
                    } else {
                        printf("\033[32m%c\033[0m", matrix[i][j]);  
                    }
                    
                    
                    if (rand() % 8 == 0) {
                        matrix[i][j] = ' ';
                    }
                } else {
                    printf(" ");
                }
            }
            printf("\n");
        }
        
        fflush(stdout);
        usleep(DELAY);
    }
    
    printf("\033[?25h");
    return 0;
}
